stan/math/opencl/prim/normal_id_glm_lpdf.hpp

```diff
@@ -159,7 +159,7 @@ return_type_t<T_alpha, T_beta, T_scale> normal_id_glm_lpdf(
           = from_matrix_cl<Dynamic, 1>(sigma_derivative_cl);
     } else {
       ops_partials.edge3_.partials_[0]
-          = (y_scaled_sq_sum - N) * as_scalar(inv_sigma);
+          = (y_scaled_sq_sum - N) * assume_type<double>(inv_sigma);
```

Comment:
The only way I see that can be done is to make each such branch into a separate function. That would make code much harder to read. Unless you have some other suggestion how to achieve that?

Reply:
Another idea how to would be to use something like:
```
ifelse<condition>([&](auto arg){
    code if true;
},[&](auto arg){
    code if false;
}
```
I think regular if and `assume_type` still makes code a bit more readable.

Reply:
> The only way I see that can be done is to make each such branch into a separate function. T

That's how we did it in the past.  I think it's a tradeoff on understandability.  It took me a bit of time to figure out what `assume_type` was doing, which is why I suggested adding doc, as it's a non-standard solution to this problem.

I just wanted to verify this is why all these `assume_type` calls were being made.  The compiler should eliminate the branches and code, so that's no big deal. 

I'd still rather see it called `forward_as`, as that indicates what it does, not why it's used.

Reply:
I am not attached to the name. To me it seem `forward_as` might be understood as if some type conversion is happening. But if you think it is clearer I will rename it.

Reply:
^I would look at `forward_as` and think it has something to do with `std::forward`